### PR TITLE
[Merged by Bors] - feat: the 'hint' tactic

### PR DESCRIPTION
**Pull request by kim-em**
```
example (h : 1 < 0) : False := by hint
example {P Q : Prop} (p : P) (f : P → Q) : Q := by hint
example {P Q R: Prop} (x : P ∧ Q ∧ R ∧ R) : Q ∧ P ∧ R := by hint
example {a b : ℚ} (h : a < b) : ¬ b < a := by hint
example : 37^2 - 35^2 = 72 * 2 := by hint
example : Nat.Prime 37 := by hint
```

Tries out any tactics registered using `register_hint tac`, and reports which ones succeed using the new "Try these: " multiple suggestion widgets. Tactics that close the goal are highlighted in green. Tactics that succeed but don't close the goal display the new subgoals in the widget. If `tac` produces a "Try this: " message, use that instead of `tac`.

I haven't hooked up `aesop` yet, because of https://github.com/leanprover-community/aesop/issues/85. Similarly for `norm_num`.

I would like to parallelize this, but I don't think that needs to happen right away.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by timotree3** [comment]
How does this compare with #5363?

**Comment by semorrison** [comment]
> How does this compare with #5363?

Oh! I had completely failed to notice that. Apologies in particular to @Komyyy.

I'm impressed how parallel parts of the implementation are. :-) My version uses the new "Try these:" widget for providing multiple suggestions, and also does some additional post-processing to stop if it finds a tactic that closes the goal, and also shows the remaining subgoals for each suggestion in the widget.

@Komyyy, how should we proceed? Can we combine these into something better than either?

**Comment by semorrison** [comment]
I will immediately steal the `let tac : Syntax.Tactic := ⟨tac.raw.copyHeadTailInfoFrom .missing⟩` trick, which I presume solves the unreachable tactic warning.

**Comment by Komyyy** [comment]
I integrated my changes:
1. Remove some placeholders for mathport
2. Remove some comments for `add_hint_tactic`
3. Align import
4. Add an new process for `register_hint` to remove following comments
5. Register the `register_hint` syntax as an exception of `UnreachableTactic` linter

**Comment by semorrison** [comment]
Thank you, @Komyyy, and apologies again for missing your earlier PR.

**Comment by semorrison** [comment]
Okay, I have a modification to this that runs all the hint tactics in parallel, and it's lovely. I still need to move some lemmas around to get the imports to work out, so we should merge this in the meantime.

File: Mathlib/Tactic/Common.lean [hunk 1]
```diff
@@ -4,14 +4,12 @@ Released under Apache 2.0 license as described in the file LICENSE.
...
-thereby making tactics widely available without needing specific imports.
+-- This file imports all tactics which do not have significant theory imports,
+-- and hence can be imported very low in the theory import hierarchy,
+-- thereby making tactics widely available without needing specific imports.
 
-We include some commented out imports here, with an explanation of their theory requirements,
-to save some time for anyone wondering why they are not here.
--/
+-- We include some commented out imports here, with an explanation of their theory requirements,
+-- to save some time for anyone wondering why they are not here.
```
Comment (kmill) [inline Mathlib/Tactic/Common.lean:12]: What's the reason for this change?

Could you revert changing the block comment to line comments, and maybe switch to `/-!` so that it shows up in the generated docs?

File: Mathlib/Tactic/Hint.lean [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import Std.Tactic.TryThis
+import Std.Linter.UnreachableTactic
+import Mathlib.Data.Nondet.Basic
+import Mathlib.Tactic.FailIfNoProgress
+import Mathlib.Mathport.Rename
+
+#align_import tactic.hint from "leanprover-community/mathlib"@"8f6fd1b69096c6a587f745d354306c0d46396915"
```
Comment (kmill) [inline Mathlib/Tactic/Hint.lean:12]: I don't think we align tactic modules, right?

File: Mathlib/Tactic/Hint.lean [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+/-- Return the list of registered hint tactics. -/
+def getHints : CoreM (List (TSyntax `tactic)) := return hintExtension.getState (← getEnv)
+
+open Lean.Elab.Command in
+/--
+Register a tactic for use with the `hint` tactic, e.g. `register_hint simp_all`.
+-/
+elab (name := registerHintStx) "register_hint" tac:tactic : command => liftTermElabM do
+  -- remove comments
+  let tac : TSyntax `tactic := ⟨tac.raw.copyHeadTailInfoFrom .missing⟩
```
Comment (kmill) [inline Mathlib/Tactic/Hint.lean:51]: I'm curious if this is the right thing to do. I know there's `tac.raw.unsetTrailing` as well, which doesn't clear source position information.

Maybe source position doesn't matter since it's not going to be used for any error reporting.

File: test/hint.lean [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import Mathlib.Tactic.Common
+import Mathlib.Tactic.Linarith
+import Mathlib.Data.Nat.Prime
+
+example (h : 1 < 0) : False := by hint
```
Comment (kmill) [inline test/hint.lean:5]: Could we use `#guard_msgs` to see what these are testing?

File: Mathlib/Tactic/Common.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/Tactic/Common.lean:12]: Sorry, yes. The linter was complaining and I forgot to revert this after finding a better solution.

Comment (semorrison) [reply Mathlib/Tactic/Common.lean:12]: It can't be `/-!` because it needs to come before the imports for readability.

Comment (semorrison) [reply Mathlib/Tactic/Common.lean]: I think it's okay to just put it after the imports, after all.

File: Mathlib/Tactic/Hint.lean (hunk 2 above)
Comment (semorrison) [reply Mathlib/Tactic/Hint.lean:12]: Yes, I agree. @Komyyy added this one, but I don't think will object to removing it.

File: test/hint.lean (hunk 4 above)
Comment (semorrison) [reply test/hint.lean:5]: Sure. I'm a little worried that once I parallelize the tactic execution the output will become nondeterministic, but we can cross that bridge later.

Comment (semorrison) [reply test/hint.lean:10]: There's also one that will change once `simp` stops calling `decide`, but again, I'll fix that later.

File: Mathlib/Tactic/Hint.lean (hunk 3 above)
Comment (semorrison) [reply Mathlib/Tactic/Hint.lean:49]: I'm not sure.

If anything, removing all source position here seems safer: it would be terribly wrong to use it!

File: Mathlib/Tactic/Common.lean [hunk 5]
```diff
@@ -112,3 +104,27 @@ import Mathlib.Util.AssertExists
...
+-/
+
+/-!
+# Register tactics with `hint`.
+-/
+
+section Hint
+
+register_hint split
+register_hint intro
```
Comment (j-loreaux) [inline Mathlib/Tactic/Common.lean:124]: Should `tauto` be included here?

File: Mathlib/Tactic/Common.lean [hunk 6]
```diff
@@ -112,3 +104,27 @@ import Mathlib.Util.AssertExists
...
+register_hint aesop
+register_hint decide
+register_hint simp_all?
+register_hint exact?
+
```
Comment (adomani) [inline Mathlib/Tactic/Common.lean:129]:
```suggestion
register_hint exact?
register_hint constructor

```
I think that Lean 3 had it and in Lean 4 is maybe more useful.
I wonder whether `split_ifs` is also an option.

File: Mathlib/Tactic/Common.lean (hunk 5 above)
Comment (semorrison) [reply Mathlib/Tactic/Common.lean:124]: At present there is no way to remove a `register_hint` (without PRing to Mathlib), but it is easy to add.

Can I suggest that intentionally undershoot on this PR, and then let people try it out and revise the list?

File: Mathlib/Tactic/Common.lean (hunk 6 above)
Comment (semorrison) [reply Mathlib/Tactic/Common.lean:129]: At present there is no way to remove a `register_hint` (without PRing to Mathlib), but it is easy to add.

Can I suggest that intentionally undershoot on this PR, and then let people try it out and revise the list?

File: scripts/lint-style.py [hunk 7]
```diff
@@ -254,7 +254,7 @@ def regular_check(lines, path):
         if copy_done and line == "\n":
             continue
         words = line.split()
-        if words[0] != "import" and words[0] != "/-!" and words[0] != "#align_import":
+        if words[0] != "import" and words[0] != "--" and words[0] != "/-!" and words[0] != "#align_import":
```
Comment (kmill) [inline scripts/lint-style.py:257]: I'm not sure this should this be part of this PR, but if you think it's OK I'll let it slide.

I think this is a good first version of the tactic, and I'm happy if you merged it.

bors d+

**Comment by semorrison** [comment]
bors merge
